doc/nom_recipes.md

```diff
@@ -0,0 +1,235 @@
+      many0(alt((alphanumeric1, tag("_"))))
+    )
+  )(input)
+}
+```
+
+Let's say we apply this to the identifier `hello_world123abc`.
+The first `alt` parser would recognize `hello_world`. The `pair` combinator ensures that `123abc` will be piped to the next `alphanumeric0` parser, which recognizes every remaining character.
+However, the `pair` combinator returns a tuple of the results of its sub-parsers. The `recognize` parser produces a `&str` of the input text that was parsed.
+
+## Literal Values
+
+### Escaped Strings
+
+This is [one of the examples](https://github.com/Geal/nom/blob/master/examples/string.rs) in the examples directory.
+
+### Integers
+
+The following recipes all return string slices rather than integer values. How to obtain an integer value instead is demonstrated for hexadecimal integers. The others are similar.
+
+The parsers allow the grouping character `_`, which allows one to group the digits by byte, for example: `0xA4_3F_11_28`. If you prefer to exclude the `_` character, the lambda to convert from a string slice to an integer value is slightly simpler. You can also strip the `_` from the string slice that is returned, which is demonstrated in the second hexdecimal number parser.
+
+If you wish to limit the number of digits in a valid integer literal, replace `many1` with `many_m_n` in the recipes: `many1(one_of("_01"))` becomes `many_m_n(1, 16, one_of("_01"))` for binary numbers having at most 16 bits.
+
+#### Hexadecimal
+
+The parser outputs the string slice of the digits without the leading `0x`/`0X`.
+
+```rust
+fn hexadecimal<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str,&'a str, E> {
+  preceded(
+    alt((tag("0x"), tag("0X"))),
+    recognize(
+      many1(
+        one_of("_0123456789abcdefABCDEF")
+      )
+    )
+  )(input)
+}
+```
+
+If you want it to return the integer value instead, use map:
+
+```rust
+fn hexadecimal_value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, i64, E> {
+  map(preceded(
+    alt((tag("0x"), tag("0X"))),
+      recognize(
+        many1(
+          one_of("_0123456789abcdefABCDEF")
+        )
+      )
+    ),
+    |out: &str| i64::from_str_radix(&str::replace(&out, "_", ""), 16).unwrap()
+  )(input)
+}
+```
+
+#### Octal
+
+```rust
+fn octal<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str,&'a str, E> {
+  preceded(
+    alt((tag("0o"), tag("0O"))),
+    recognize(
+      many1(
+        one_of("_01234567")
+      )
+    )
+  )(input)
+}
+```
+
+#### Binary
+
+```rust
+fn binary<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str,&'a str, E> {
+  preceded(
+    alt((tag("0b"), tag("0B"))),
+    recognize(
+      many1(
+        one_of("_01")
+      )
+    )
+  )(input)
+}
+```
+
+#### Decimal
+
+```rust
+fn decimal<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str,&'a str, E> {
+  recognize(
+    tuple((
+      one_of("0123456789"),
+      many0(one_of("_0123456789")
+      )
```

Comment:
Nit: this parenthesis should be on the previous line

Reply:
Ach! Makes me twitch.